zml/module.zig

```diff
@@ -639,46 +646,51 @@ pub const CompilationContext = struct {
         defer arena_state.deinit();
+        // This arena is used for allocations which won't outlive the function call,
+        // but the function creation uses `self.allocator()` which we'll live for the duration of the compilation.
         const arena = arena_state.allocator();
 
         // first, do the "compile" and check the bytecode
         // the result of this will also have the correct tags of the result shapes
         const args_hash = hashArgs(args);
-        const key: FnCache.Key = .{ .fn_ptr = &func, .input_hash = args_hash };
+        const key: FnKey = .{ .fn_ptr = &func, .input_hash = args_hash };
 
-        const function = self._fn_cache.getEntry(key) orelse b: {
+        const function = self._fn_cache.get(key) orelse b: {
             const full_name: [:0]const u8 = if (std.mem.eql(u8, "main", func_name))
-                arena.dupeZ(u8, func_name) catch unreachable
+                try self.allocator().dupeZ(u8, func_name)
             else
-                std.fmt.allocPrintZ(arena, "{s}_{x}", .{ func_name, key.input_hash }) catch unreachable;
+                try std.fmt.allocPrintZ(self.allocator(), "{s}_{x}", .{ func_name, key.input_hash });
 
             var arg_id: u16 = 0;
             var tensor_args: @TypeOf(args) = args;
-            meta.mapAlloc(struct {
+            try meta.mapAlloc(struct {
                 fn cb(arg_id_: *u16, x: Tensor) Tensor {
                     const a = arg_id_.*;
                     arg_id_.* += 1;
                     return Tensor{ ._shape = x._shape, ._id = .{ .arg_id = a }, ._donation = .{ .arg = a } };
                 }
-            }.cb, arena, &arg_id, args, &tensor_args) catch @panic("OutOfMemory");
+            }.cb, arena, &arg_id, args, &tensor_args);
 
-            const f = self.emitMlir(func, &tensor_args, .{
-                .name = full_name,
-            }) catch @panic("OOM");
+            const f = try self.emitMlir(
+                func,
+                &tensor_args,
+                .{ .name = full_name },
+            );
             self._module.getBody().appendOperation(f.mlir_fn);
 
-            break :b self._fn_cache.addEntry(self.allocator(), key, f) catch unreachable;
+            try self._fn_cache.putNoClobber(self.allocator(), key, f);
+            break :b f;
         };
 
         const loc = self.mlirCtx().location(@src());
 
-        const values = arena.alloc(mlir.Value, function.num_args) catch unreachable;
+        const values = try arena.alloc(mlir.Value, function.num_args);
         self.extractValues(&args, values);
 
-        const donations = arena.alloc(Tensor._Donation, function.num_args) catch unreachable;
+        const donations = try arena.alloc(Tensor._Donation, function.num_args);
         meta.collectBuf(struct {
             pub fn cb(ctx: *const CompilationContext, x: Tensor) Tensor._Donation {
                 return ctx.getValueAndDonation(x)[1];
```

Comment:
We should probably clean that at some point because I see a lot of potential leaks

Reply:
I've doubled check the logic, allocations are split in two: memory for the caller which is part of the result and memory which is temporary because copied by MLIR.

Reply:
The fact that we use `self.allocator()` instead of `self._arena.allocator()` makes that unclear without having the context. But yes, it seems it's okay